opm/simulators/utils/compressPartition.hpp

```diff
@@ -0,0 +1,33 @@
+/*
+  Copyright 2023 Equinor ASA
+
+  This file is part of the Open Porous Media Project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_UTIL_COMPRESS_PARTITION_HPP_INCLUDED
+#define OPM_UTIL_COMPRESS_PARTITION_HPP_INCLUDED
+
+#include <utility>
+#include <vector>
+
+namespace Opm { namespace util {
+    std::pair<std::vector<int>, int>
+    compressAndCountPartitionIDs(std::vector<int>&& parts0);
```

Comment:
Perhaps early to comment as this is still a draft but anyway: I would prefer that these functions took an lvalue reference instead, as that would make it possible to use them in-place. As is, if I have a partition vector `p` I'd like to compress I will have to move it to this function, than afterwards move the result to the original `p`.

Reply:
> I would prefer that these functions took an lvalue reference instead, as that would make it possible to use them in-place.

I don't think it makes sense to have this particular compression function work in-place.  In addition to compressing, this function also counts the number of blocks&ndash;i.e., the number of unique block IDs&ndash;and it returns a pair  of the compressed partition vector and that number of unique block IDs.  We could add an overload of the other function, `void compressPartitionIDs(std::vector<int>& parts)`, though.  I'd rather not have a function which both mutates an in/out parameter **and** returns an integer value.